### ISSUE TITLE
Dynamic Glossary collation

### ISSUE THREAD

**Issue by dbolack-ab** [naturalcrit/homebrewery#3370]
### Your idea:

Dynamic Glossary collation follows the pattern set up in Rich Indexes and definition lists.

Arbitrarily placed markdown is placed in the brew text and then collated and ordered by a snippet which outputs any number of glossary lists.

Markdown Format:
The markdown takes one of two shapes, each as a single line entry.
1. `# Glossary Term :: Glossary Definition` - This creates a term/definition pair and places it on the default glossary list `Glossary`
2. `#GlossaryList Glossary Term :: Glossary  Definition` - This creates a term/definition pair and places it on the user-created  glossary list `GlossaryList`

Markdown details:
In both cases, the markdown is consumed on parsing.

Snippet Details:
The snippet runs its own parsing of the markdown from the brew and places it in the designated Glossary List. 
Duplicate terms will have their Definitions collected for shared output.  
Each list will be ordered by term. 
Output for each list will have the list name as an H2 header, followed by the list rendered as a PanDoc style <DL> list followed by a `\page`

**Comment by ericscheid** [comment]
Looks good.

I assume that the presence of the double-colon `::` is what distinguishes this syntax from the index-marking syntax of #3369?

Slight suggestion might be to use single colon to terminate the glossary-list-name, like this
`#Glossary List Name: Term :: Glossary Definition`

**Comment by dbolack-ab** [comment]
> I assume that the presence of the double-colon `::` is what distinguishes this syntax from the index-marking syntax of #3369?

Correct.

> Slight suggestion might be to use single column to terminate the glossary-list-name, like this `#Glossary List Name: Term :: Glossary Definition`

This is a good idea, though I think I want to want and see how the discussion shakes out on putting the list name at the end in 3369 and use the same pattern in both places.

**Comment by ericscheid** [comment]
Revisiting this ... there are two forms of the <this-is-a-glossary-entry> snippet (one with a default assumed glossary name, and a more complex form with a glossary name).

Ordinarily this would mean two snippets. I'm OK with that — a simple comparison of what the two snippets insert would also serve to document/explain how the 2nd snippet names the glossary.

What's missing though would be a _third_ snippet, one that collates all the glossary entries and writes them out with the glossary name as a heading on a new page. And a UI mechanism to select which of many glossaries to do this for[1].

---- 

1. If I have multiple glossaries, I wouldn't want one action to rewrite multiple pages .. I might have additional markup added to make each glossary page distinct and pretty e.g. insert an appropriate image onto the "Herbs" glossary page.

**Comment by ericscheid** [comment]
Commenting on #4152 ...

> which syntax should be used for what? Prepare the terms vs display the list?

If I'm reading the PR right, the two snippets are both for displaying [i.e. collating] the list. There is no snippet for inserting the syntax for marking a text as being a glossary item.

I note also the syntax is no longer demarcated with `::`, but instead with `//`. 

How are authors meant to know they need to enter `term // definition`? Are they meant to intuit somehow that they should use the syntax from definition-terms, but to use the separator from the insert-index-marker syntax (i.e. `\\`)?

Also, for ease of editing, perhaps the default/simple syntax could be the usual definition-list-item .. but with a `#Glossary: ` prefix. This would make it very easy to convert from in-line definition-list-item on-the-page as needed, and convert to shunt-to-glossary syntax (and back again). 

That is, restore the "default" glossary name back into the base syntax. This does away with the need for two snippets (for two syntax forms), as well as make it very easy for an author to rename a glossary as needed (e.g. "#Glossary:" → "#Herbs:", or "#Glossary:" → "#Słowniczek:")

### In summary:

1. one snippet that inserts `#Glossary: term :: definition`
2. one snippet for generating the glossary  
(`\page\n\n{{glossary\n# Glossary\n<collated-definition-list-entries>\n}}\n`
3. if there is more than 1 glossary name detected, pop a little dialog for choosing which particular glossary to insert (or replace?)
4. editing the literal default string of "Glossary" changes the glossary the entry will appear in
5.  removing the `#<name>: ` prefix converts the glossary entry back to in-page definition-term syntax
6. adding the `#<name>: ` prefix converts the in-page definition-term syntax to glossary-entry syntax

**Comment by dbolack-ab** [comment]
> Ordinarily this would mean two snippets. I'm OK with that — a simple comparison of what the two snippets insert would also serve to document/explain how the 2nd snippet names the glossary.

There is no reason for distinct snippets. 

> What's missing though would be a _third_ snippet, one that collates all the glossary entries and writes them out with the glossary name as a heading on a new page. And a UI mechanism to select which of many glossaries to do this for[1].

I'm not adding a UI layer.  It is absolutely unnecessary.

> 1. If I have multiple glossaries, I wouldn't want one action to rewrite multiple pages .. I might have additional markup added to make each glossary page distinct and pretty e.g. insert an appropriate image onto the "Herbs" glossary page.

That would better handled by repouring and pasting across the versions.

**Comment by dbolack-ab** [comment]
> I note also the syntax is no longer demarcated with `::`, but instead with `//`.

Yes.

> 
> How are authors meant to know they need to enter `term // definition`? Are they meant to intuit somehow that they should use the syntax from definition-terms, but to use the separator from the insert-index-marker syntax (i.e. `\\`)?

Documentation.

> That is, restore the "default" glossary name back into the base syntax. This does away with the need for two snippets (for two syntax forms), as well as make it very easy for an author to rename a glossary as needed (e.g. "#Glossary:" → "#Herbs:", or "#Glossary:" → "#Słowniczek:")

I prefer the brevity and would prefer to not potentially confuse users between this and the DD syntax. this does not seem very Markdownish.

> 
> ### In summary:
> 1. one snippet that inserts `#Glossary: term :: definition`
> 2. one snippet for generating the glossary
>    (`\page\n\n{{glossary\n# Glossary\n<collated-definition-list-entries>\n}}\n`
> 3. if there is more than 1 glossary name detected, pop a little dialog for choosing which particular glossary to insert (or replace?)
> 4. editing the literal default string of "Glossary" changes the glossary the entry will appear in
> 5. removing the `#<name>: ` prefix converts the glossary entry back to in-page definition-term syntax
> 6. adding the `#<name>: ` prefix converts the in-page definition-term syntax to glossary-entry syntax

Not a thing there I agree with.

**Comment by ericscheid** [comment]
> Not a thing there I agree with.

Well, you do agree that editing the `<glossary-name>` part of the entry syntax changes which glossary it appears in.

Lemme unpack some of the nuances I was assuming and keep the discussion moving forward

- any proposal can be split into MVP and then afterwards enhanced

- the reason the syntax should follow the defn-list syntax (and not the indexing syntax) is that the glossary syntax is for declaring a term and it's definition, and the defn-list syntax is for declaring a term and it's definition — they differ only in where that defn & term appears in the document; and neither is for defining a landmark position in the document.

- though documentation can be helpful it should not be the first resort. Firstly, lack of documentation is a recognised weakness of this project (one which should be addressed, of course). Secondly, it is a well established principle in ergonomics and UX that slapping a label or instruction on a thing does not fix usability issues, and assuming the user should look elsewhere for that instruction is user hostile.


### In summary (expanded):

#### in mvp
1. two snippets  
• snippet "Glossary Entry" that inserts the syntax for an individual glossary entry  
• snippet "Generate Glossary" for generating the glossary
2. only one syntax for a glossary entry  
• uses `#<glossary-name>:` prefix as it's signature  
• uses the default glossary name "Glossary"
• is based on the definition-list syntax  
3. editing the `<glossary-name>` part of the syntax changes which glossary the entry will appear in
4. when generating the glossary: insert at editing cursor, insert all glossaries found

#### future enhancement
5. if there is more than 1 glossary name detected: supply UI for choosing which particular glossary to insert  
• this could be via a popup dialog,  
• or via multiple instances of "Generate Glossary → glossary name" on the snippet menu  
• (or some other UI mechanism)
6. allow colons in the glossary name via the escaping-syntax-to-text standard (e.g. `#glossary\:with name: <term and definition>`)
7. the "Glossary Entry" snippet inserts the appropriate translation of the "Glossary" string for brews in other languages.

#### ease of use for authors
8. not rely on users reading documentation to discover basic syntax
9. removing the `#<name>: ` prefix converts the glossary entry to in-page definition-term syntax
10. adding the `#<name>: ` prefix converts the in-page definition-term syntax to glossary-entry syntax

#### note
glossary entry syntax as `#<glossary-name><optional-whitespace>:<optional-whitespace><term> :: <defn>`
- starts on a new line
- no space after the initial `#` (to differentiate from headings)
- `<term-and-defn>` uses `<definition-list-entry>` syntax  
for ease of editing between "inline in page" and "collate to glossary" use cases

**Comment by dbolack-ab** [comment]
> > Not a thing there I agree with.
> Well, you do agree that editing the `<glossary-name>` part of the entry syntax changes which glossary it appears in.

That is how I wrote it, so I suppose it is reasonable to expect I agree with 

> * any proposal can be split into MVP and then afterwards enhanced

With this project, I don't think so. This needs to land intended feature complete.

> * the reason the syntax should follow the defn-list syntax (and not the indexing syntax) is that the glossary syntax is for declaring a term and it's definition, and the defn-list syntax is for declaring a term and it's definition — they differ only in where that defn & term appears in the document; and neither is for defining a landmark position in the document.

I agree with this principle; however, one of the core problems with marked is the lack of precision control over token parsing. We've particularly had problems with `:` combos due to DL list construction. This is the main reason I swapped this out. 

> should look elsewhere for that instruction is user hostile.

Documentation is now anti-user? Lord.

> 1. two snippets
>    • snippet "Glossary Entry" that inserts the syntax for an individual glossary entry
>    • snippet "Generate Glossary" for generating the glossary
> 2. only one syntax for a glossary entry

No.

> 6. allow colons in the glossary name via the escaping-syntax-to-text standard (e.g. `#glossary\:with name: <term and definition>`)

Should be there from the start, no matter the delimiter.

> 9. removing the `#<name>: ` prefix converts the glossary entry to in-page definition-term syntax

I mean... that's what happens in the text editor....

**Comment by ericscheid** [comment]
> > should look elsewhere for that instruction is user hostile.
> 
> Documentation is now anti-user? Lord.

Documentation is good. *Hiding* that documentation is anti-user.

**Comment by ericscheid** [comment]
> No.

Explain.
